kong2kic/types.go

```diff
@@ -0,0 +1,259 @@
+package kong2kic
+
+import (
+	"encoding/json"
+
+	"github.com/kong/go-database-reconciler/pkg/file"
+	kicv1 "github.com/kong/kubernetes-ingress-controller/v3/pkg/apis/configuration/v1"
+	kicv1beta1 "github.com/kong/kubernetes-ingress-controller/v3/pkg/apis/configuration/v1beta1"
+	k8scorev1 "k8s.io/api/core/v1"
+	k8snetv1 "k8s.io/api/networking/v1"
+	k8sgwapiv1 "sigs.k8s.io/gateway-api/apis/v1"
+	"sigs.k8s.io/yaml"
+)
+
+// KICContent represents a serialized Kong state for KIC.
+// +k8s:deepcopy-gen=true
+type KICContent struct {
+	KongIngresses        []kicv1.KongIngress             `json:"kongIngresses,omitempty" yaml:",omitempty"`
+	KongPlugins          []kicv1.KongPlugin              `json:"kongPlugins,omitempty" yaml:",omitempty"`
+	KongClusterPlugins   []kicv1.KongClusterPlugin       `json:"clusterPlugins,omitempty" yaml:",omitempty"`
+	Ingresses            []k8snetv1.Ingress              `json:"ingresses,omitempty" yaml:",omitempty"`
+	Services             []k8scorev1.Service             `json:"services,omitempty" yaml:",omitempty"`
+	Secrets              []k8scorev1.Secret              `json:"secrets,omitempty" yaml:",omitempty"`
+	KongConsumers        []kicv1.KongConsumer            `json:"consumers,omitempty" yaml:",omitempty"`
+	KongConsumerGroups   []kicv1beta1.KongConsumerGroup  `json:"consumerGroups,omitempty" yaml:",omitempty"`
+	HTTPRoutes           []k8sgwapiv1.HTTPRoute          `json:"httpRoutes,omitempty" yaml:",omitempty"`
+	KongUpstreamPolicies []kicv1beta1.KongUpstreamPolicy `json:"upstreamPolicies,omitempty" yaml:",omitempty"`
+}
+
+func (k KICContent) marshalKICContentToYaml() ([]byte, error) {
```

Comment:
are `marshalKICContentToJSON` and `marshalKICContentToYaml` the same except for the format used? If so, we should consolidate them into a single one:

```
func (k KICContent) marshalKICContentToFormat(format string) ([]byte, error) {
	var output []byte

	const (
		yamlSeparator = "---\n"
	)

	for _, kongIngress := range k.KongIngresses {
		kongIngresses, err := SerializeObjectDroppingFields(kongIngress, format)
		if err != nil {
			return nil, err
		}
		output = append(output, kongIngresses...)
		if format == file.YAML {
			output = append(output, []byte(yamlSeparator)...)
		}
	}
...
...
}
```